### ISSUE TITLE
Incorrect behavior on maps with referenced attribute values

### ISSUE THREAD

**Issue by michalmedvecky** [hashicorp/terraform#18908]
When adding attributes to maps, with values referencing to other TF objects, it evaluates to an empty map.
 
### Terraform Version
```
0.11.7
```

### Terraform Configuration Files
```hcl
locals {
  mtdt = {
    groups = "asdf"
#       ip = "${openstack_networking_port_v2.test.all_fixed_ips.0}"
  }
}

#resource "openstack_networking_port_v2" "test" {
#  name               = "port"
#  network_id         = "d3ebd619-3f2d-45cb-8e28-86da7cdf4cc5"
#  admin_state_up     = "true"
#}

resource "openstack_compute_instance_v2" "test" {
  name = "test"
  image_id = "66ae409c-f2ea-4a8c-a25c-3859d79f9f33"
  flavor_name = "m1.small"
  key_pair = "misko"
  availability_zone = "mz"
  metadata = "${local.mtdt}"
  network {
    name = "sandbox"
  }
}
```

### Steps to Reproduce
0. add ^ code to `test.tf`
1. `terraform init`
2. `terraform apply`
3. `vim test.tf`, uncomment all commented lines
4. `terraform plan`

### Expected Behavior

TF should add one metadata item with a computed value.

```
  ~ openstack_compute_instance_v2.test
      metadata.%:               "" => <computed>
      metadata.ip:              "" => "<computed>"
```

### Actual Behavior

TF removes all actual metadata

```
  ~ openstack_compute_instance_v2.test
      metadata.%:               "" => <computed>
      metadata.groups:          "asdf" => ""
```

**Comment by apparentlymart** [comment]
Hi @michalmedvecky! Sorry for this strange behavior, and thanks for reporting it.

Over the last few months we've been working to finish up a set of changes to the configuration language engine to fix weird issues like this and add some new features. That work is now merged into master, so I wanted to check what you reported to see if it is fixed in the v0.12.0-alpha2 prerelease build.

I started with the following simple config that mimics yours but only uses local resources, for easier test environment setup:

```hcl
locals {
  x = {
    foo = "bar"
  }
}

resource "null_resource" "a" {
  triggers = locals.x
}
```

```
$ terraform apply

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # null_resource.a will be created
  + resource "null_resource" "a" {
      + id       = (known after apply)
      + triggers = {
          + "foo" = "bar"
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

null_resource.a: Creating...
null_resource.a: Creation complete after 0s [id=6093501951867186545]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

I then added a new resource and added it to the map similar to what you did:

```hcl
locals {
  x = {
    foo = "bar"
    bar = null_resource.b.id
  }
}

resource "null_resource" "a" {
  triggers = local.x
}

resource "null_resource" "b" {
}
```

```
$ terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

null_resource.a: Refreshing state... [id=6093501951867186545]

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # null_resource.a must be replaced
-/+ resource "null_resource" "a" {
      ~ id       = "6093501951867186545" -> (known after apply)
      ~ triggers = {
          - "foo" = "bar"
        } -> (known after apply)
    }

  # null_resource.b will be created
  + resource "null_resource" "b" {
      + id = (known after apply)
    }

Plan: 2 to add, 0 to change, 1 to destroy.
```

Looking at this rendering produced the new diff renderer we've integrated recently, I can see the cause of what you saw: because the new id isn't known yet, Terraform is marking _the entire map_ as computed, and so it doesn't know the new keys. The old diff renderer presented this in a very confusing way, marking the `metadata.%` field as changing to computed. The new diff renderer is more clear about what it's intending to do:

```
      ~ triggers = {
          - "foo" = "bar"
        } -> (known after apply)
```

...but it's still strange that the whole map would become unknown, rather than just the new element:

```
      ~ triggers = {
            "foo" = "bar"
          + "bar" = (known after apply)
        }
```

... so it does seem like there is something remaining to be fixed here.

With all of that said, in your case I think in spite of this incorrect planned change the end result would still have _worked_. Once Terraform had created your `openstack_networking_port_v2.test` during the apply step it would compute the final value of that map just in time to update `openstack_compute_instance_v2.test` to include both keys in  `metadata`. We can see this in my example by accepting the proposed change and then looking at the resulting state of `null_resource.a`:

```
$ terraform apply
null_resource.a: Refreshing state... [id=6093501951867186545]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # null_resource.a must be replaced
-/+ resource "null_resource" "a" {
      ~ id       = "6093501951867186545" -> (known after apply)
      ~ triggers = {
          - "foo" = "bar"
        } -> (known after apply)
    }

  # null_resource.b will be created
  + resource "null_resource" "b" {
      + id = (known after apply)
    }

Plan: 2 to add, 0 to change, 1 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

null_resource.a: Destroying... [id=6093501951867186545]
null_resource.a: Destruction complete after 0s
null_resource.b: Creating...
null_resource.b: Creation complete after 0s [id=6648669433891428174]
null_resource.a: Creating...
null_resource.a: Creation complete after 0s [id=6548312243722228153]

Apply complete! Resources: 2 added, 1 changed, 0 destroyed.

$ terraform show
# null_resource.a: 
resource "null_resource" "a" {
    id = "6548312243722228153"
    triggers = {
        "bar" = "6648669433891428174"
        "foo" = "bar"
    }
}

# null_resource.b: 
resource "null_resource" "b" {
    id = "6648669433891428174"
}
```

Since this seems to be mainly a cosmetic/UI issue, we'll postpone this for a later release just because the 0.12.0 release scope is already very large and we're now in the prerelease phase and so trying to keep further changes to a minimum to reduce risk. However, we will leave this open and look into what's going on here in a future release.

Thanks again for reporting this!

**Comment by mildwonkey** [comment]
Hi @michalmedvecky !
It looks like this has been fixed in terraform 0.12, possibly as a side-effect of some other bug or enhancement!

To use martin's example, adding the second `null_resource` (after applying the example code) and running `apply` no longer reports an unknown value in the first resource:


```
$ terraform apply
null_resource.a: Refreshing state... [id=4135111821422281842]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # null_resource.b will be created
  + resource "null_resource" "b" {
      + id = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

null_resource.b: Creating...
null_resource.b: Creation complete after 0s [id=593531844485388508]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

Thanks for reporting this! I am going to close this issue now.
